bundles/core/org.eclipse.smarthome.core.semantics/src/main/java/org/eclipse/smarthome/core/semantics/internal/SemanticsMetadataProvider.java

```diff
@@ -0,0 +1,266 @@
+
+    // the namespace to use for the metadata
+    public static final String NAMESPACE = "semantics";
+
+    // holds the static definition of the relations between entities
+    private static Map<List<Class<? extends Tag>>, String> RELATIONS_PARENT = new HashMap<>();
+    private static Map<List<Class<? extends Tag>>, String> RELATIONS_MEMBER = new HashMap<>();
+    private static Map<List<Class<? extends Tag>>, String> RELATIONS_PROPERTY = new HashMap<>();
+
+    static {
+        RELATIONS_PARENT.put(Arrays.asList(Equipment.class, Location.class), "hasLocation");
+        RELATIONS_PARENT.put(Arrays.asList(Point.class, Location.class), "hasLocation");
+        RELATIONS_PARENT.put(Arrays.asList(Location.class, Location.class), "isPartOf");
+        RELATIONS_PARENT.put(Arrays.asList(Equipment.class, Equipment.class), "isPartOf");
+        RELATIONS_PARENT.put(Arrays.asList(Point.class, Equipment.class), "isPointOf");
+
+        RELATIONS_MEMBER.put(Arrays.asList(Equipment.class, Point.class), "hasPoint");
+
+        RELATIONS_PROPERTY.put(Arrays.asList(Point.class), "relatesTo");
+    }
+
+    // local cache of the created metadata as a map from itemName->Metadata
+    private final Map<String, Metadata> semantics = new TreeMap<>(new Comparator<String>() {
+        @Override
+        public int compare(String s1, String s2) {
+            return s1.compareTo(s2);
+        }
+    });
+
+    @NonNullByDefault({})
+    private ItemRegistry itemRegistry;
+
+    @NonNullByDefault({})
+    private TagService tagService;
+
+    @Activate
+    protected void activate() {
+        for (Item item : itemRegistry.getAll()) {
+            processItem(item);
+        }
+        itemRegistry.addRegistryChangeListener(this);
+    }
+
+    protected void deactivate() {
+        itemRegistry.removeRegistryChangeListener(this);
+        semantics.clear();
+
+    }
+
+    @Reference
+    protected void setItemRegistry(ItemRegistry itemRegistry) {
+        this.itemRegistry = itemRegistry;
+    }
+
+    protected void unsetItemRegistry(ItemRegistry itemRegistry) {
+        this.itemRegistry = null;
+    }
+
+    @Reference
+    protected void setTagService(TagService tagService) {
+        this.tagService = tagService;
+    }
+
+    protected void unsetTagService(TagService tagService) {
+        this.tagService = null;
+    }
+
+    @Override
+    public Collection<Metadata> getAll() {
+        return semantics.values();
+    }
+
+    /**
+     * Updates the semantic metadata for an item and notifies all listeners about changes
+     *
+     * @param item the item to update the metadata for
+     */
+    @SuppressWarnings({ "null", "unused" })
+    private void processItem(Item item) {
+        MetadataKey key = new MetadataKey(NAMESPACE, item.getName());
+        Map<String, Object> configuration = new HashMap<>();
+        Class<? extends Tag> type = tagService.getSemanticType(item);
+        if (type != null) {
```

Comment:
This doesn't process the hierarchy unless a tag is directly applied on the item, so it doesn't handle situations like this:
```
Group gGF           "Ground Floor"  <groundfloor> ["Floor"]
Group:Switch:OR(ON, OFF)        Lights      "All Lights [(%d)]"                 <lightbulb> ["Lightbulb"]
Switch Light_GF_Corridor_Ceiling    "Downstairs Corridor"       (GF_Corridor, Lights)
```
In this case I would expect something like this:
```
    "link": "http://localhost:8080/rest/items/Light_GF_Corridor_Ceiling",
    "metadata": {
      "semantics": {
        "value": "Equipment_Lightbulb", // not actually sure how it would come up with it - prone to conflicts
        "config": {
          "isPartOf": "Lights",
          "hasLocation": "GF_Corridor"
        }
      }
    },
```

Reply:
You will notice that I purposefully didn't add ANY tags to the functional group items (Lights, Heatings, etc.) in my example above. I think this can easily get us into trouble. It would actually mean "This item is a lightbulb", while you actually intend to express "All items in this group are lightbulbs". I think it is the safer way to require all items to be individually tagged accordingly.
MAYBE we can do an exception for Point (i.e. real functionality) types, because they will never(?) be assigned to a group item. So if they are found on a group item, once could declare that this means that it should be applied to all members.

Reply:
> I think it is the safer way to require all items to be individually tagged accordingly.

I'm still not 100% happy with this from a user convenience perspective :) ...but I recall a prior conversation about assuming groups as an entity by itself vs. groups as a set of similar items so I understand.
Not sure about introducing too many exceptions, it would just add to the confusion.